compose: Migrate to a shared compose banner template with shared styling.

evykassirer opened the pull request:
Part one of fixing #22524.

Updated warning banners:

* sending to a resolved topic
* person tagged is not subscribed
* sending a wildcard that will notify more than 15 people
* naming a private stream in a message to a public stream

Updated error banners:

* sending an empty message (only for ctrl+enter to send)
* using wildcards in a channel where they're not allowed
* "Error checking subscription"
* sending to a stream that doesn't exist
* no permission to post there
* private messages are disabled
* missing private message recipient
* recipient is invalid (seems to only happen with cross-realm emails, or somehow entering recipients that don't exist even though we prevent you from doing that)
* recipients (plural) are invalid
* message is too long
* topic field is empty (only if empty topic isn't allowed)
* stream field is empty
* sending to a stream you're not subscribed to
* "You cannot send messages to deactivated users."
* "You need to be running Zephyr mirroring in order to send messages!"

Notable pieces that are **not** part of this PR:
* upload banner
* success messages
* moving banners outside of the compose box

**Screenshots and screen captures:**

![image](https://user-images.githubusercontent.com/5634097/197048641-e97373aa-c1c5-40ea-8df6-967e5204d004.png)

<img width="902" alt="image" src="https://user-images.githubusercontent.com/5634097/191351071-f65c15da-81c2-4323-b18a-1db64aff2452.png">
<img width="893" alt="image" src="https://user-images.githubusercontent.com/5634097/191351108-970d4daa-d6d4-468d-bb70-6c14b56c8d44.png">
<img width="873" alt="image" src="https://user-images.githubusercontent.com/5634097/191351142-c474a7f1-5761-4854-a5ba-eaea9c2db068.png">
<img width="893" alt="image" src="https://user-images.githubusercontent.com/5634097/191351192-e82213b4-abf6-47dd-b6c7-9c29af99a279.png">
<img width="885" alt="image" src="https://user-images.githubusercontent.com/5634097/191351227-7863328d-a9bd-4dd3-82ba-4609a3adfce7.png">




**Self-review checklist**

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [x] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [x] Explains differences from previous plans (e.g., issue description).
- [x] Highlights technical choices and bugs encountered.
- [x] Calls out remaining decisions and concerns.
- [x] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/version-control.html)).

- [x] Each commit is a coherent idea.
- [x] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [x] Visual appearance of the changes.
- [x] Responsiveness and internationalization.
- [x] Strings and tooltips.
- [x] End-to-end functionality of buttons, interactions and flows.
- [x] Corner cases, error conditions, and easily imagined bugs.

timabbott commented on static/templates/compose_banner.hbs:
```diff
@@ -0,0 +1,7 @@
+<div class="compose_banner {{banner_type}}" id="{{css_id}}" data-stream-id="{{stream_id}}" data-topic-name="{{topic_name}}">
```
Maybe `element_id` would be better than `css_id`, since it's technically an HTML element ID, not a CSS thing.

timabbott commented on static/js/compose_validate.js:
```diff
@@ -191,27 +194,35 @@ export function warn_if_topic_resolved(topic_changed) {
+            css_id: "compose_resolved_topic",
         };
 
-        const new_row = render_compose_resolved_topic(context);
```
I think you missed deleting the HTML template for this, which is helpful to look at, as it reveals where the strings come from.

timabbott commented on static/js/compose_error.ts:
```diff
@@ -1,7 +1,39 @@
+
 import * as common from "./common";
 
+// TODO: where should these live? don't want a circular import. ask: where do constants usually go?
```
Best place is in the module that owns them, which probably should be `compose_error`.

evykassirer commented on static/js/compose_error.ts:
```diff
@@ -1,27 +1,83 @@
 import $ from "jquery";
```
I wonder if this file should be called `compose_banner.ts`, since that seems like a more descriptive name now

timabbott replied:
I think that'd be a reasonable name change; if you want to do that, we should do the rename in its own commit; maybe we can decide when this branch is merged and just do the rename as a follow-up if we like the idea.

timabbott commented on tools/lib/template_parser.py:
```diff
@@ -138,8 +138,8 @@ def looking_at_whitespace() -> bool:
-                tag = s[9:-2]
-                kind = "handlebars_singleton"
+                tag = s[5:-2]
+                kind = "handlebars_partial"
```
If it's easy to add a test in `tools/tests/test_template_parser.py`, that's an important part of how I'd make changes to this library, since @showell is our expert on it.

Also, are you sure this doesn't break anything else? We do use the `{{>` syntax elsewhere, and I'm kinda guessing that you should be adding a new block, not changing that existing one.

evykassirer replied:
It's confusing to me that the old one also called it a partial, but I decided not to look into that for now and just rename the old one to `singleton` more consistently. Also added tests.

timabbott replied:
How did you pick these names? It looks like the thing we already have is a "partial" and the thing you're adding is a "partial block", so maybe we should use those names?

https://handlebarsjs.com/guide/partials.html#partial-blocks

timabbott replied:
I think the original `handlebars_singleton` name is the thing that's confusing -- can you rename it to be consistently `handlebars_partial` and the new one to `handlebars_partial_block`? Then we're using Handlebars upstream terminology for these two concepts, rather than whatever was made up when working on this tool.

timabbott commented on static/templates/stream_does_not_exist_error.hbs:
```diff
@@ -0,0 +1,9 @@
+{{#> compose_banner }}
```
Is it easy to put these special error message templates in a `static/templates/compose_banner/` directory? We've been trying to clear the root directory of this sort of error message template, and I think that would save us from wanting to name this `compose_banner_stream_does_not...` for clarity.

evykassirer replied:
that sounds great! I was also not very happy with the `compose_` prefix here

timabbott commented on static/js/compose_error.ts:
```diff
@@ -1,11 +1,68 @@
+    $("#compose-send-button").removeClass("disable-btn");
+}
+
+// NOTE: if you add a new classname, make sure it's added to ALL_BANNER_MESSAGES above.
```
What happens if it isn't? I think it's reasonable to have that case throw a `blueslip.error`.

The other direction we could consider is having `ERROR_MESSAGES` be a map, and the caller is supposed to pass, for example, `compose_banner.ERROR_MESSAGES.invalid_recipient`; then it'd be hard to write code that didn't pass a value in that object.

timabbott commented on static/js/compose_error.ts:
```diff
@@ -1,11 +1,68 @@
+export function show_error_message(message: string, classname: string, $bad_input?: JQuery): void {
+    if (classname === "generic_compose_error" && $(".generic_compose_error").length) {
+        // Generic errors overwrite each other, so if one already exists we just update the text.
+        $(".generic_compose_error .banner_content").text(message);
```
Does this block need to do anything to clear non-text content that may have been inserted by other banner types?

timabbott replied:
Also, what's the thinking behind having this block instead of just rerendering the thing from scratch whenever requested? I'm not sure in what situations having a special case here would improve things.

evykassirer replied:
> Does this block need to do anything to clear non-text content that may have been inserted by other banner types?

I was wondering about that, and felt mostly confident that it would override that, but not totally sure.

> Also, what's the thinking behind having this block instead of just rerendering the thing from scratch whenever requested? I'm not sure in what situations having a special case here would improve things.

You're totally right, I was stuck in the pattern I was copying from. This should just clear the existing generic message and continue adding a banner normally like the others. That would be way better and address the first issue

timabbott commented on static/js/compose.js:
```diff
@@ -516,17 +510,6 @@ export function initialize() {
-        if ($all_invites.children().length === 0) {
-            $all_invites.hide();
-        }
-    });
```
What did this code do -- and does the new implementation do this somewhere else?

evykassirer replied:
Previously the way it worked was there would be a parent div for each kind of banner (the ones I removed from `compose.hbs`) and banners would be put inside the relevant div when created.

Now the way it works is there is a single div that contains all banners.

This deleted code checks to see if there are no more warnings of this type, and hides the parent div if so. This isn't relevant to the new way it works because there are no parent divs that are show/hidden.

If that doesn't make sense I can try to clarify further.

timabbott commented on static/templates/compose_all_everyone_warning.hbs:
```diff
@@ -0,0 +1,9 @@
+    <br />
+    If you don't want to do that, please edit your message to remove the <strong>@{wildcard_mention}</strong> mention.
+    {{/tr}}
+{{/compose_banner}}
```
I don't know what commit we want to put this in, but we should rename it to something like `compose_banner/wildcard_warning.hbs`? The "all_everyone" label is a very silly way to refer to "wildcard" mentions :).

evykassirer replied:
Ah yeah, I think this is the only error message I didn't change the name of to be something more reasonable, because I forgot we call them wildcards! It's a new file, so I think choosing a good name is a good idea.

timabbott commented:
I posted a fairly detailed review of this; hopefully that offers options for how to resolve the remaining details. I think skipping the uploading banners for now is a very reasonable simplification, though I'd manually test to make sure they aren't affected by this.

I don't really understand the issue with `X` styling in the last comment; probably something to summarize in #design or #frontend for Vlad to help resolve.

alya commented:
Will uploading banner be a short-term follow-up? We really need to fix the fact that the current coloring looks totally out of place in dark theme.

evykassirer commented:
Yeah! I was planning on continuing on this until all the compose banners were done, and _then_ doing the PR that moves it out of the composebox. I just wanted to cut this PR off at some reasonable point so that it didn't get too unwieldy. I can also have both PRs ready to go before merging them if you'd rather merge them at the same time.

timabbott commented:
Yeah, I agree with the strategy of not trying to address uploading banners in this PR; it has a lot of complexity that we've already reviewed pretty extensively, and it'll be much easier to integrate that as a follow-up.

andersk commented on static/styles/compose.css:
```diff
@@ -276,51 +276,96 @@
-    margin-right: -2px;
-    width: 10px;
-}
+    .action_button {
```
We do not want to be moving in this direction of unqualified CSS class names. They hinder modularity, even when “scoped” using nesting, because the CSS notion of scope doesn’t match with the conceptual notion of scope in components. See

https://chat.zulip.org/#narrow/stream/6-frontend/topic/CSS.20nesting/near/1283355

alya commented:
Playing with this briefly:

1. Has dark theme not been implemented yet? I see:

![Screen Shot 2022-09-14 at 3 55 31 PM](https://user-images.githubusercontent.com/2090066/190277213-1943b98f-03ca-4a31-a8e6-70695d3660cf.png)

2. Also, in the above screenshot, the spacing between "Subscribe them" and `x` looks oddly wide to me.

alya commented:
3. When the text goes to multiple lines, the vertical spacing at the top and bottom looks too small to me. @terpimost what was your intention for this?
![Screen Shot 2022-09-14 at 3 57 16 PM](https://user-images.githubusercontent.com/2090066/190277459-ea4dffab-7471-4275-bd04-8ba6676a652b.png)

evykassirer commented on frontend_tests/node_tests/compose.js:
```diff
@@ -109,7 +108,16 @@ function initialize_handlers({override}) {
+    // zjquery doesn't support `remove`, which is used when clearing the compose box.
+    // TODO: improve how we test this so that we don't have to mock things like this.
+    for (const classname of Object.values(compose_banner.CLASSNAMES)) {
+        $(`#compose_banners .${classname}`).remove = noop;
```
There are a lot of places where I do something like `#compose_banners .${classname}` and I'm not a huge fan of that `.` being outside of the classname, but I ended up doing this because I wanted to pass `compose_banner.CLASSNAMES.____` directly as the classname to the template, where it shouldn't have the dot. I'm curious if others have thoughts on ways to do this more nicely.

evykassirer commented on frontend_tests/node_tests/compose_actions.js:
```diff
@@ -67,6 +68,17 @@ const respond_to_message = compose_actions.respond_to_message;
 
+// TODO: dedupe this function --- ASK where a good place for shared tooling like this is,
+// since I don't see many shared functions in unit tests. Or is it better to just have
+// duplicated code like this?
```
question for a reviewer!

evykassirer commented on frontend_tests/node_tests/compose_validate.js:
```diff
@@ -460,72 +519,69 @@ test_ui("test_validate_stream_message_post_policy_full_members_only", () => {
     assert.ok(!$textarea.hasClass("over_limit"));
-});
-
-test_ui("test_message_overflow", () => {
```
I couldn't see how this test tested anything relevant that the above test wasn't already testing.

timabbott replied:
Seems reasonable.

terpimost commented:
About the vertical spacing. I'm expecting the padding from edges be the same no matter the length of the content. Also please don't forget about the proper line height.
<img width="909" alt="image" src="https://user-images.githubusercontent.com/1903309/190312404-92e73a63-d734-4bab-a212-8f798b9908be.png">

<img width="154" alt="image" src="https://user-images.githubusercontent.com/1903309/190312577-2ed70fe3-a68e-4e08-a849-51c80a9a80f7.png">
I think border radius is off, I made them 4px, but in the example it looks like 10px...

terpimost commented:
I understand that this PR and the issue might be not about that aspect of banner. But the main concern was to not have extra `X` button if the alert is inside the message box. Current alerts should be visually outside the compose box...

terpimost commented:
It looks like red color in the light theme is off:
<img width="1182" alt="image" src="https://user-images.githubusercontent.com/1903309/190941979-688ad6fb-b1c3-4ba6-9b2a-9a08873e0c9d.png">

also i think close icon is too bold, did you take it [from here](https://www.figma.com/file/jbNOiBWvbtLuHaiTj4CW0G/Zulip-Web-App?node-id=2857%3A208867)?

alya commented:
A couple of comments from playing with the PR:

1. I noticed that some banners have different margin on the left than others, which seems odd:

<img width="907" alt="Screen Shot 2022-09-20 at 4 38 12 PM" src="https://user-images.githubusercontent.com/2090066/191383595-d080e0d8-519e-4163-a59f-7e71306272cb.png">

2. Something about this banner feels off to me.  I think the weirdness is having a larger vertical margin on the bottom than the top.

<img width="902" alt="Screen Shot 2022-09-20 at 4 40 48 PM" src="https://user-images.githubusercontent.com/2090066/191383620-7030ea14-737d-4a18-9457-2034f7710279.png">

Thanks for continuing to iterate on this!

evykassirer commented:
Those all look like old banners! I'm not sure why you're not seeing the new ones

terpimost commented:
Hight is a bit bigger than should be:
<img width="628" alt="image" src="https://user-images.githubusercontent.com/1903309/191408302-a0a5bcc0-b93e-40f4-b621-2397f9703bbe.png">

Close button is too far from the right edge and it is too small. Close button rectangle should be bigger:
<img width="278" alt="image" src="https://user-images.githubusercontent.com/1903309/191408461-de1000aa-b8c1-4882-bf42-365ff4978a8b.png">

alya commented:
> Those all look like old banners! I'm not sure why you're not seeing the new ones. Two of the ones you're screenshotting are in my screenshots above 🤔

Hmm, let me try again!

alya commented:
Hmm, testing this again, I'm not seeing the warning for mentioning a private stream.

alya commented:
Otherwise looks good in my non-comprehensive manual testing. :)

evykassirer commented:
Oh I should be more specific about that since I also got confused when testing this -- it only works if you name the private stream directly in a non-private stream (e.g. #kandra), not if you reference any particular message. I think we should probably change that, but that's how it works in CZO right now as well. Does that fix the issue you saw, or were you unable to reproduce something that worked on CZO?

alya commented:
I just named a private stream directly. Let me know if you retest this and can't replicate.

evykassirer commented:
<img width="779" alt="image" src="https://user-images.githubusercontent.com/5634097/192584480-b83aefb9-a9e4-4e6a-a53d-fbabaac780cd.png">

alya commented:
No warning for #core team for me here:

![2022-09-27 09 54 23](https://user-images.githubusercontent.com/2090066/192588678-690d589b-a135-4cea-add9-5ee493295787.gif)

evykassirer commented:
Ah, I think the issue is you need to have someone subscribed to #private that isn't in #core team

alya commented:
Ahh, Zulip is so smart!

OK, I think @timabbott can take a look next, and we'll also need approval from @terpimost before we test-deploy.

timabbott commented on static/js/compose_validate.js:
```diff
@@ -191,29 +189,34 @@ export function warn_if_topic_resolved(topic_changed) {
             return;
         }
 
+        const button_text = settings_data.user_can_move_messages_between_streams()
```
This is a preexisting issue, so should get its own commit to fix, but we should be checking `user_can_edit_topic_of_any_message`, not `user_can_move_messages_between_streams`, not resolving a topic does not involve moving messages between streams.

evykassirer replied:
done

timabbott commented on static/templates/compose_banner/compose_banner.hbs:
```diff
@@ -0,0 +1,7 @@
+<div class="compose_banner {{banner_type}} {{classname}}" data-stream-id="{{stream_id}}" data-topic-name="{{topic_name}}">
```
The `data-stream-id` and `data-topic-name` fields seem like they should be inside a conditional -- won't this result in something like `data-stream-id="undefined"` for a banner appearing attached to a private message conversation...

evykassirer replied:
Got confused about how to do this and asked on [CZO](https://chat.zulip.org/#narrow/stream/6-frontend/topic/conditional.20data.20in.20template/near/1469288)

evykassirer replied:
ok done :)

timabbott replied:
Looks great!

timabbott commented on static/js/hotkey.js:
```diff
@@ -308,6 +308,10 @@ export function process_escape_key(e) {
 
+            if ($("#compose_banners").length) {
+                $("#compose_banners").empty();
+            }
```
What's the thinking behind adding this here? We don't have a similar check for the preexisting banners system.

Might make sense to put this in a separate commit with a commit message explaining the change; and also it's probably worth having a comment here for posterity with some color for why we want to clear compose banners.

Though I think it's probably a bug to have this here and `compose_actions.cancel` or one of its subroutines should be responsible?

evykassirer replied:
We do a similar check for the preexisting banners system right above this.

```
            if ($("#compose-send-status").css("display") !== "none") {
                $("#compose-send-status").hide();
                return true;
            }
```

timabbott replied:
That other version returns `true` -- which means it considers this the complete processing of the `Esc` keyboard shortcut (i.e. it'll just hide the banner, and not also close the compose box). Shouldn't this one do that? 

(And I guess we haven't ported 100% of things, so we don't get to delete the old one).

evykassirer replied:
Yup, the old one gets deleted in the upcoming Upload Banner PR, once we've stopped using `$("#compose-send-status")` completely :)

And yes good catch, I'll add the `return true`

timabbott replied:
The `return true` got placed one commit after the code introducing this check, which seems like a rebase mistake. I've fixed that in my local copy but FYI, the best way to catch this sort of thing is to just look at the commits in `gitk` or a similar viewer after doing the rebase to check your work.

timabbott commented on static/js/reminder.js:
```diff
@@ -96,7 +99,7 @@ export function schedule_message(request = compose.create_message_object()) {
     const error = function (response) {
         $("#compose-textarea").prop("disabled", false);
-        compose_error.show(_.escape(response), $("#compose-textarea"));
+        compose_error.show_error_message(response, $("#compose-textarea"));
```
This seems like a bug; the second parameter is supposed to be a string...

evykassirer replied:
thanks!

timabbott commented on frontend_tests/node_tests/compose_validate.js:
```diff
@@ -711,93 +719,39 @@ test_ui("warn_if_mentioning_unsubscribed_user", ({override, override_rewire, moc
+    new_banner_rendered = false;
+    $("#compose_banners .recipient_not_subscribed").length = 0;
+    compose_validate.warn_if_mentioning_unsubscribed_user(mentioned_user);
+    assert.ok(new_banner_rendered);
```
This test is no longer validating any of the details of the warning; just a boolean for whether something was rendered. 

I think the right structure for this class of test would be to just verify an exact object we expect to have been passed to the template; I guess the old one did that in some form. I'm not sure it's worth addressing in this PR.

evykassirer replied:
It tests the details of the warning in the template mock, in the way you're describing, above:

```
    mock_template("compose_banner/not_subscribed_warning.hbs", false, (data) => {
        assert.equal(data.classname, compose_error.CLASSNAMES.recipient_not_subscribed);
        assert.equal(data.user_id, 34);
        assert.equal(data.stream_id, 111);
        assert.equal(data.name, "Foo Barson");
        new_banner_rendered = true;
    });
```

timabbott replied:
OK sounds good; had missed that because a lot of similar looking things were removed; this test was pretty messy :).

evykassirer replied:
hopefully much cleaner now!

timabbott commented on static/js/compose.js:
```diff
@@ -495,53 +494,43 @@ export function initialize() {
-                compose_error.CLASSNAMES.generic_compose_error,
-                $("#compose-textarea"),
+            const $invite_row = $(event.target).parents(
+                `.${compose_error.CLASSNAMES.recipient_not_subscribed}`,
```
This could be just `.parents("compose_banner")`; might be worth doing just to avoid unnecessary references to the CLASSNAMES lookup.

evykassirer replied:
sounds good, and also changed in other relevant place

evykassirer commented on frontend_tests/node_tests/compose_validate.js:
```diff
@@ -673,22 +672,30 @@ test_ui("warn_if_private_stream_is_linked", ({mock_template}) => {
     override(settings_data, "user_can_subscribe_other_users", () => true);
 
-    let mentioned = {
+    let mentioned_user = {
```
The name made sense to me because the line `people.add_active_user(mentioned_user);`, but I can change it to `mention_details` if you think that is clearer.

timabbott replied:
Yeah, I think that's better -- generally a thing that says "user" should be some version of a user object.

evykassirer commented on static/js/zcommand.js:
```diff
@@ -50,12 +48,7 @@ export function send(opts) {
-        .stop(true)
-        .fadeTo(0, 1);
-    $("#compose-error-msg").text(msg);
+    compose_error.show_error_message(msg, "generic_compose_error");
```
yeah good catch, I'm not sure why that happened 🙃

timabbott replied:
It seems like this one ended up as `generic_error` in `reminder.js` and `generic_compose_error` in the other places.

I'll fix this. But it's a little sad that `compose_banner` being TypeScript isn't sufficient for TypeScript to catch this for us.

timabbott commented:
I fixed a few things noted above and pushed back here.

The one bug/regression I noticed is that these notices should probably be singletons:

![image](https://user-images.githubusercontent.com/2746074/204872828-5b23b8dc-1c31-4dd1-9961-e40a73a3438e.png)

![image](https://user-images.githubusercontent.com/2746074/204873315-52b5ecb0-63a2-4d4b-a520-bff1edb50d3b.png)

I haven't tested whether other alerts have this bug as well. (It's correct that the notice for mentioning someone on another stream is not). I'm not sure what the right rule is -- do we want to hide an existing alert with the same class name and `data-stream-id` (at least for the private stream case) when trying to display a new one? Or do we want to try to avoid adding the new one if it's already there (keeping in mind that "duplicates" for different streams are probably correct).

I plan to test-deploy this while you address that, so we can get feedback on the design change.

evykassirer commented:
I agree there are some issues there -- I think there was a Zulip thread about that somewhere, and I thought it was in the issue description but I can't find it now.

Essentially some of these aren't regressions but existing issues, and we should probably have separate GitHub issues to track fixing them

<img width="881" alt="image" src="https://user-images.githubusercontent.com/5634097/204877821-30293c55-a8af-4f77-a2e2-77e37b0aa776.png"> -- this one is tricky because, as you mentioned, it's possible that they mention multiple private streams in the same message. We can do something to account for this, but it feels out of scope here.

The "not subscribed" seems like a regression though, I can fix that. I can go through and see which ones do duplicates and which should be fixed in the future.

evykassirer commented:
✅ `render_compose_banner` -- in every use, only one banner of a classname can be visible at a time (one with `user_not_subscribed` wasn't doing this before, which looks like the only case of this regression)
✅  `render_not_subscribed_warning` -- does not duplicate names if an unsubscribed user is mentioned multiple times
✅ `render_wildcard_warning` -- only one of this banner is shown at a time
✅ `render_stream_does_not_exist_error` -- only one of this banner is shown at a time

---

Noted followup tasks (none of these work currently):

* `private_stream_warning` shouldn't show multiple times for the same stream
* `user_not_subscribed` should go away if the composebox stream switches to a stream that the user is subscribed to
* `stream_does_not_exist_error` should go away if the composebox stream switches to a valid stream
*  `recipient_not_subscribed` doesn't go away if the user's name is removed, but that seems maybe more difficult to fix than it's worth

There might be more cases for when banners should go away, but I didn't get that into that right now.

timabbott commented:
> recipient_not_subscribed doesn't go away if the user's name is removed, but that seems maybe more difficult to fix than it's worth

This sort of thing can be tricky since theoretically it should matter if the name is inside a code block; and also we just don't want to do anything expensive every time you type a character.

evykassirer commented:
Yeah that's what I was imagining. I think leaving that one as is seems okay for now

timabbott commented:
Test-deploying this on chat.zulip.org.

One corner case that I don't think we've discussed is whether these buttons should be right-aligned; I think the current implementation is fine but a consequence of the left alignment is this:

![image](https://user-images.githubusercontent.com/2746074/205193005-8db68507-bb33-43ae-8d75-cb49fbea967b.png)

evykassirer commented:
Changes since last time:

* fixed button alignment (as discussed on [CZO](https://chat.zulip.org/#narrow/stream/101-design/topic/UI.20redesign.3A.20compose.20warnings/near/1472957))

timabbott commented on static/js/reminder.js:
```diff
@@ -96,7 +99,11 @@ export function schedule_message(request = compose.create_message_object()) {
-        compose_error.show(_.escape(response), $("#compose-textarea"));
+        compose_error.show_error_message(
+            response,
+            compose_error.CLASSNAMES.generic_error,
```
This also appears broken again.

evykassirer replied:
Oh sorry! Not part of my usual workflow. I'm also happy to add the fixes back if that's easier for you.

timabbott replied:
Nah, I've already pulled it down and done the rebase.

evykassirer commented:
Yay!
